### PR TITLE
[go_router_builder] Support default values for route parameters

### PR DESCRIPTION
**Pull request by ValentinVignal**
Closes https://github.com/flutter/flutter/issues/114406

This PR adds the support of default values in the route constructors.
If the URL/location doesn't have a set value for the field, it uses the default value. If the value of the field is equal to the default value, it doesn't include it in the location/URL.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [relevant style guides] and ran the auto-formatter. (Unlike the flutter/flutter repo, the flutter/packages repo does use `dart format`.)
- [x] I signed the [CLA].
- [x] The title of the PR starts with the name of the package surrounded by square brackets, e.g. `[shared_preferences]`
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated `pubspec.yaml` with an appropriate new version according to the [pub versioning philosophy], or this PR is [exempt from version changes].
- [x] I updated `CHANGELOG.md` to add a description of the change, [following repository CHANGELOG style].
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[relevant style guides]: https://github.com/flutter/packages/blob/main/CONTRIBUTING.md#style
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[pub versioning philosophy]: https://dart.dev/tools/pub/versioning
[exempt from version changes]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#version-and-changelog-updates
[following repository CHANGELOG style]: https://github.com/flutter/flutter/wiki/Contributing-to-Plugins-and-Packages#changelog-style
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests

### CODE REVIEW

File: packages/go_router_builder/lib/src/route_config.dart [hunk 1]
```diff
@@ -315,11 +315,15 @@ GoRouteData.\$route(
 
     final StringBuffer buffer = StringBuffer('queryParams: {\n');
 
-    for (final String param
-        in _ctorQueryParams.map((ParameterElement e) => e.name)) {
+    for (final ParameterElement param in _ctorQueryParams) {
+      final String parameterName = param.name;
+      String condition = '$parameterName != null';
+      if (param.hasDefaultValue) {
+        condition += '&& $parameterName != ${param.defaultValueCode!}';
```
Comment (chunhtai) [inline packages/go_router_builder/lib/src/route_config.dart:322]: How do you tell the difference between a parameter with default value but explicitly set to null? in that case the URL it generate will always represents parameter with default value

Comment (ValentinVignal) [reply packages/go_router_builder/lib/src/route_config.dart:322]: I was considering the type being non-nullable when a default value is set. But you are right, this is not always true. I assume we could allow / need `?param=null` in the URL

Comment (chunhtai) [reply packages/go_router_builder/lib/src/route_config.dart:322]: yes, I can't think of other way to support nullable field without som caveats. Only allow non-nullable field to have default value SGTM

Comment (ValentinVignal) [reply packages/go_router_builder/lib/src/route_config.dart:322]: So what do you think about the current implementation? Do you want me to change something?

**Comment by ValentinVignal** [comment]
@chunhtai Is there anything else you want me to do about this?

File: packages/go_router_builder/lib/src/route_config.dart (hunk 1 above)
Comment (chunhtai) [reply packages/go_router_builder/lib/src/route_config.dart:322]: The more I think about this the more I think we probably shouldn't add this api. cc @johnpryan WDYT

Comment (ValentinVignal) [reply packages/go_router_builder/lib/src/route_config.dart:322]: Maybe I can add some more insights:

I have to have a route that has a lot of parameters that are:
- No nullable
- Optional
- With a default value

For now, I have to do:

```dart
@TypedGoRoute<MyRoute>(path: '/my-route')
class MyRoute extends GoRouteData {
  const MyRoute({
    String? paramString,
    int? paramInt,
    // ... A lot more
  }): paramString = paramString ?? 'defaultValue',
    paramInt = paramInt ?? 0
    // ... A lot more
    ;

  final String paramString;
  final int paramInt;
  // ... A lot more
} 
```

The issue is that the URL becomes quite lengthy and not user-friendly.

If I do

```dart
context.go(const MyRoute().location);
```

The URL will be `/my-route?paramString=defaultValue&paramInt=0` and even longer with the other parameters.
I would like my URL to be `/my-route` only

Comment (chunhtai) [reply packages/go_router_builder/lib/src/route_config.dart:322]: you can handle the default value in build method.

If we really want to do this, we need some kind of warning or even linting error if developer assign null to the nullable parameter.

Comment (ValentinVignal) [reply packages/go_router_builder/lib/src/route_config.dart:322]: > If we really want to do this, we need some kind of warning or even linting error if developer assign null to the nullable parameter.

But didn't we say the parameter would be non-nullable? (https://github.com/flutter/packages/pull/2826#discussion_r1043635704)

Comment (chunhtai) [reply packages/go_router_builder/lib/src/route_config.dart:322]: yes but what i meant was that we need something to warn developer about it.

Comment (ValentinVignal) [reply packages/go_router_builder/lib/src/route_config.dart:322]: I can try to implement a warning or an error if the type is nullable

Comment (ValentinVignal) [reply packages/go_router_builder/lib/src/route_config.dart]: What do you think about [🔊 Throw an error when a default value is used with a nullable type](https://github.com/flutter/packages/pull/2826/commits/3a9432f16c8f10cf673b1508060fc61902a28db7)?

Now it throws an error if a default value is used with a nullable type

**Comment by stuartmorgan** [comment]
@chunhtai Ping from triage: what's the status of this review?

**Review by chunhtai — COMMENTED** [review]
sorry for the late reply, just come back from vacation. I think this looks good now. Can you add update the readme to include this new functionality?

**Comment by ValentinVignal** [comment]
@chunhtai I've modified the README in [Update README](https://github.com/flutter/packages/pull/2826/commits/3946cc71dba2a44ce6f056b4894438b75921351f). Tell me if that's good enough for you

**Review by chunhtai — APPROVED** [review]
LGTM

File: packages/go_router_builder/README.md [hunk 2]
```diff
@@ -179,6 +179,23 @@ class LoginRoute extends GoRouteData {
...
+class MyRoute extends GoRouteData {
+  MyRoute({this.queryParameter = 'defaultValue'});
+  final String from;
+
+  @override
+  Widget build(BuildContext context, GoRouterState state) => MyScreen(from: from);
+}
+```
+
+A query parameter equal to its default value is not included in the location.
```
[the remark this answers is not in the document]
Comment (ValentinVignal) [reply packages/go_router_builder/README.md:196]: Actually _"equal"_ is not the verb in the sentence, the verb is _"is"_

_"A query parameter equals to its default value"_ is the subject of the sentence

[the remark this answers is not in the document]
Comment (chunhtai) [reply packages/go_router_builder/README.md:196]: ah sorry, should it be this then? "A query parameter that equals to its default value"

[the remark this answers is not in the document]
Comment (ValentinVignal) [reply packages/go_router_builder/README.md]: Sure, I've changed it

**Review by hangyujin — APPROVED** [review]
LGTM

**Pull request merged**